### PR TITLE
Prac 07 feedback

### PR DESCRIPTION
**Pull request by JevanChandra**
@AdriantoHartono
@a-rosa
@MikeLeongZhiFai
@Riston-gitting-gud

Please help review my code

### CODE REVIEW

File: cp1404practicals/prac_07/README.md
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/README.md:1]: Nice, and good written 👌

File: cp1404practicals/prac_07/guitars.csv
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/guitars.csv:1]: Great👌

File: cp1404practicals/prac_07/language_file_reader.py
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/language_file_reader.py:1]: Nice code, and already written nicely👌

File: cp1404practicals/prac_07/languages.csv
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/languages.csv:1]: Great🙏

File: cp1404practicals/prac_07/myguitars.py
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/myguitars.py:1]: Nice code, and written nicely 🙏

File: cp1404practicals/prac_07/prac_06_guitar.py
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/prac_06_guitar.py:1]: Great class code👌

File: cp1404practicals/prac_07/programming_language.py
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/programming_language.py:1]: Great code👌

File: cp1404practicals/prac_07/project.py
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/project.py:1]: Great class code👍

File: cp1404practicals/prac_07/project_management.py
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/project_management.py:1]: Great code, it is already nicely written in each function👍

File: cp1404practicals/prac_07/projects.txt
Comment (AdriantoHartono) [inline cp1404practicals/prac_07/projects.txt:1]: Great🙏

**Comment by MikeLeongZhiFai** [comment]
All the codes are working well!

File: cp1404practicals/prac_07/language_file_reader.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        # print(repr(line))  # debugging
+        # Strip newline from end and split it into parts (CSV)
+        parts = line.strip().split(',')
+        # print(parts)  # debugging
+        # Reflection is stored as a string (Yes/No) and we want a Boolean
+        reflection = parts[2] == "Yes"
+        pointer_arithmetic = parts[4] == "Yes"
+        # Construct a ProgrammingLanguage object using the elements
+        # year should be an int
+        language = ProgrammingLanguage(parts[0], parts[1], reflection, int(parts[3]), pointer_arithmetic)
```
Comment (a-rosa) [inline cp1404practicals/prac_07/language_file_reader.py:32]: You still need the year on the end of the list, you are needed to add pointer_arithmetic not replace year with it

File: cp1404practicals/prac_07/language_file_reader.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+# using_csv()
+
+
+def using_namedtuple():
+    """Language file reader version using a named tuple."""
+    in_file = open('languages.csv', 'r', newline='')
+    file_field_names = in_file.readline().strip().split(',')
+    print(file_field_names)
+    # Language will be a new subclass of the tuple data type class
+    Language = namedtuple('Language', 'name, typing, reflection, year')
```
Comment (a-rosa) [inline cp1404practicals/prac_07/language_file_reader.py:67]: add pointer arithmetic in the tuple

File: cp1404practicals/prac_07/language_file_reader.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        print(repr(language))
+    in_file.close()
+
+
+# using_namedtuple()
+
+
+def using_csv_namedtuple():
+    """Language file reader version using both csv module and named tuple."""
+    Language = namedtuple('Language', 'name, typing, reflection, year')
```
Comment (a-rosa) [inline cp1404practicals/prac_07/language_file_reader.py:82]: add pointer arithmetics

File: cp1404practicals/prac_07/myguitars.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        print(guitar)
+
+
+def put_data_in_file(data, guitars, in_file):
+    for name, year, cost in data:
+        guitars.append(Guitar(name, int(year), float(cost)))
+    in_file.close()
+
+
+main()
```
Comment (a-rosa) [inline cp1404practicals/prac_07/myguitars.py:22]: Missing dosctring

File: cp1404practicals/prac_07/prac_06_guitar.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        return 2022 - self.year
+
+    def is_vintage(self):
+        if self.get_age() >= 50:
+            return True
+        else:
+            return False
+
+    def __lt__(self, other):
+        return self.year < other.year
```
Comment (a-rosa) [inline cp1404practicals/prac_07/prac_06_guitar.py:20]: Missing docstring

File: cp1404practicals/prac_07/programming_language.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        return f"{self.name}, {self.typing} Typing, Reflection={self.reflection}, First appeared in {self.year}, " \
+               f"Pointer Arithmetic={self.pointer_arithmetic} "
+
+    def is_dynamic(self):
+        """Determine if language is dynamically typed."""
+        return self.typing == "Dynamic"
+
+
+def run_tests():
+    """Run simple tests/demos on ProgrammingLanguage class."""
```
Comment (a-rosa) [inline cp1404practicals/prac_07/programming_language.py:29]: You didnt add pointer arithmetic part on this to run the test

File: cp1404practicals/prac_07/project.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+    def __lt__(self, other):
+        return self.priority < other.priority
+
+    def __str__(self):
+        return f"{self.name}, start: {self.start_date}, priority {self.priority}, estimate: {self.cost_estimate}, " \
+               f"completion: {self.completion_percentage}%"
+
+    def is_completed(self):
+        return self.completion_percentage == 100
```
Comment (a-rosa) [inline cp1404practicals/prac_07/project.py:17]: Missing docstring

File: cp1404practicals/prac_07/project_management.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+def save_data(filename, projects):
+    with open(f"{filename}.txt", "w") as out_file:
+        print("Name   Start Date    Priority    Cost    Estimate   Completion   Percentage", file=out_file)
+        for project in projects:
+            print(f"{project.name}   {project.start_date}   {project.priority}   {project.cost_estimate}"
+                  f"   {project.completion_percentage}", file=out_file)
+
+
+main()
```
Comment (a-rosa) [inline cp1404practicals/prac_07/project_management.py:245]: Missing docstring

Comment (Riston-gitting-gud) [inline cp1404practicals/prac_07/project_management.py:245]: Readable code and functional too! But you could remove the main() in line 234

**Pull request merged**